vms/platformvm/txs/mempool/mempool.go

```diff
@@ -5,42 +5,291 @@ package mempool
+	m.consumedUTXOs.Put(tx.TxID, tx.InputIDs())
+	m.droppedTxIDs.Evict(tx.TxID)
+	m.currentGas += gasUsed
+
+	m.updateMetrics()
+
+	return nil
+}
+
+// Evict a tx if it is lower paying and the resulting total gas is within
+// capacity
+func (m *Mempool) tryEvictTx(next gas.Gas, tx heapTx) error {
+	_, low, ok := m.minHeap.Peek()
+	if !ok {
+		return ErrGasCapacityExceeded
+	}
+
+	// Check if evicting the lowest paying tx would give us enough capacity to fit
+	// this tx
+	if low.GasPrice < tx.GasPrice && next-low.gasUsed <= m.gasCapacity {
+		// Check to see if this conflicts with any remaining txs after we evict
+		evictedUTXOs, _ := m.consumedUTXOs.DeleteKey(low.TxID)
+		if m.consumedUTXOs.HasOverlap(tx.InputIDs()) {
+			m.consumedUTXOs.Put(low.TxID, evictedUTXOs)
+			return mempool.ErrConflictsWithOtherTx
+		}
+
+		evictedTxID, _, _ := m.minHeap.Pop()
+		m.remove(evictedTxID)
+
+		return nil
+	}
+
+	return ErrGasCapacityExceeded
+}
+
+func (m *Mempool) updateMetrics() {
+	m.numTxsMetric.Set(float64(m.maxHeap.Len()))
+	m.gasAvailableMetric.Set(float64(m.gasCapacity - m.currentGas))
+}
+
+func (m *Mempool) Get(txID ids.ID) (*txs.Tx, bool) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	tx, ok := m.maxHeap.Get(txID)
+	if !ok {
+		return nil, false
+	}
+
+	return tx.Tx, true
+}
+
+func (m *Mempool) Remove(txID ids.ID) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	m.remove(txID)
+}
+
+func (m *Mempool) remove(txID ids.ID) {
+	heapTx, ok := m.maxHeap.Remove(txID)
+	if !ok {
+		return
+	}
+
+	m.minHeap.Remove(txID)
+	m.consumedUTXOs.DeleteKey(txID)
+
+	m.currentGas -= heapTx.gasUsed
+
+	m.updateMetrics()
+}
+
+func (m *Mempool) RemoveConflicts(utxos set.Set[ids.ID]) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	for _, removed := range m.consumedUTXOs.DeleteOverlapping(utxos) {
+		m.remove(removed.Key)
+	}
+}
+
+func (m *Mempool) Peek() (*txs.Tx, bool) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	_, tx, ok := m.maxHeap.Peek()
+	return tx.Tx, ok
+}
+
+func (m *Mempool) Iterate(f func(tx *txs.Tx) bool) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	for _, v := range m.maxHeap.UnsortedIterator() {
+		if !f(v.Tx) {
+			return
+		}
+	}
+}
+
+func (m *Mempool) MarkDropped(txID ids.ID, reason error) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	if _, ok := m.maxHeap.Get(txID); ok {
+		return
+	}
+
+	m.droppedTxIDs.Put(txID, reason)
+}
+
+func (m *Mempool) GetDropReason(txID ids.ID) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	err, _ := m.droppedTxIDs.Get(txID)
+	return err
+}
+
+func (m *Mempool) Len() int {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	return m.maxHeap.Len()
+}
+
+func (m *Mempool) meter(tx txs.UnsignedTx) (gas.Dimensions, gas.Gas, error) {
+	c, err := fee.TxComplexity(tx)
+	if err != nil {
+		return gas.Dimensions{}, 0, err
+	}
+
+	g, err := c.ToGas(m.weights)
+	if err != nil {
+		return gas.Dimensions{}, 0, err
+	}
+
+	if g == 0 {
+		return gas.Dimensions{}, 0, errNoGasUsed
 	}
 
-	return m.Mempool.Add(tx)
+	return c, g, nil
```

Comment:
I don't think we ever actually use `c` from here. Should we just return the gas consumed (and maybe then rename the method to be a bit more descriptive?)

Or is this going to be used in a later change of the block building logic?